Content.Server/GameObjects/EntitySystems/Atmos/GasTileOverlaySystem.cs

```diff
@@ -0,0 +1,350 @@
+                    chunk.Update(data, invalid);
+                }
+            }
+
+            var currentTime = _gameTiming.CurTime;
+
+            // Set the LastUpdate for chunks.
+            foreach (var (chunk, _) in updatedTiles)
+            {
+                chunk.Dirty(currentTime);
+            }
+
+            // Now we'll go through each player, then through each chunk in range of that player checking if the player is still in range
+            // If they are, check if they need the new data to send (i.e. if there's an overlay for the gas).
+            // Afterwards we reset all the chunk data for the next time we tick.
+            foreach (var (session, overlay) in _knownPlayerChunks)
+            {
+                if (session.AttachedEntity == null) continue;
+                
+                // Get chunks in range and update if we've moved around or the chunks have new overlay data
+                var chunksInRange = GetChunksInRange(session.AttachedEntity).ToArray();
+                var knownChunks = overlay.GetKnownChunks().ToArray();
+                var chunksToRemove = new List<GasOverlayChunk>(0);
+                var chunksToAdd = new List<GasOverlayChunk>(0);
+                
+                foreach (var chunk in chunksInRange)
+                {
+                    if (!knownChunks.Contains(chunk))
+                    {
+                        chunksToAdd.Add(chunk);
+                    }
+                }
+
+                foreach (var chunk in knownChunks)
+                {
+                    if (!chunksInRange.Contains(chunk))
+                    {
+                        chunksToRemove.Add(chunk);
+                    }
+                }
+                
+                foreach (var chunk in chunksToAdd)
+                {
+                    var message = overlay.AddChunk(currentTime, chunk);
+                    if (message != null)
+                    {
+                        RaiseNetworkEvent(message, session.ConnectedClient);
+                    }
+                }
+
+                foreach (var chunk in chunksToRemove)
+                {
+                    overlay.RemoveChunk(chunk);
+                }
+                
+                var clientInvalids = new Dictionary<GridId, List<(MapIndices, GasOverlayData)>>();
+
+                // Check for any dirty chunks in range and bundle the data to send to the client.
+                foreach (var chunk in chunksInRange)
+                {
+                    if (!updatedTiles.TryGetValue(chunk, out var invalids)) continue;
+
+                    if (!clientInvalids.TryGetValue(chunk.GridIndices, out var existingData))
+                    {
+                        existingData = new List<(MapIndices, GasOverlayData)>();
+                        clientInvalids[chunk.GridIndices] = existingData;
+                    }
+                    
+                    existingData.AddRange(chunk.GetData(invalids));
```

Comment:
You can make the `GetData` method take in the `existingData` list here so that it can directly insert into it, a bit faster.